### DISCUSSION TITLE
Taxi Cost formula?

### DISCUSSION THREAD

**Discussion by Zintixx** [AlanMorel/MapleServer2#296] [General]
We know so far that the taxi costs are already calculated in the client side. We need to figure out what the formula is.

Here's a handful of data that sharr from Discord was able to grab.:
https://docs.google.com/spreadsheets/d/12YKsVO3qMdVRUoRMvUQRBo0DaOuhVFchy-y569CLWTY/edit#gid=724092846
(2nd sheet)

it contains level, map origin, map destination, and price

It is likely that the distance and size (from world map) is taken into consideration. You'll find the coords on the right hand side.

Anyone have any idea what the formula could be?

#86

**Comment by kOchirasu** [comment]
Seems to be calculated by some lua function that takes 2 args. I'm not skilled enough to trace the call, but I have verified that the function produces the correct values using CheatEngine.

```c
// #STR: "calcTaxiCharge"
int __thiscall sub_5AA590(void *this, int a2, unsigned __int16 a3)
{
  int v3; // esi
  void (__thiscall *v4)(int); // eax
  struct lua_State *v5; // edx
  int v6; // edi
  int v7; // edi
  int v9; // edi
  void (__thiscall *v10)(int); // edx
  char v11[4]; // [esp+18h] [ebp-18h] BYREF
  int v12; // [esp+1Ch] [ebp-14h]
  int v13; // [esp+20h] [ebp-10h]
  int v14; // [esp+2Ch] [ebp-4h]

  (*(*this + 16))(this, v11);
  v14 = 0;
  if ( !v12 )
    return 0;
  v3 = v12 + 24;
  v4 = *(*(v12 + 24) + 4);
  v13 = v12 + 24;
  v4(v12 + 24);
  v5 = *(v12 + 20);
  LOBYTE(v14) = 1;
  lua_getfield(v5, -10002, aCalctaxicharge);
  v6 = v12;
  lua_pushnumber(*(v12 + 20), a2);
  ++*(v6 + 16);
  v7 = v12;
  lua_pushnumber(*(v12 + 20), a3);
  ++*(v7 + 16);
  if ( !sub_5A8780(v12) )
  {
    LOBYTE(v14) = 0;
    (*(*v3 + 8))(v3);
    v14 = -1;
    sub_A7D640(v11);
    return 0;
  }
  v9 = sub_5A8510(1);
  v10 = *(*v3 + 8);
  LOBYTE(v14) = 0;
  v10(v3);
  v14 = -1;
  sub_A7D640(v11);
  return v9;
}
```

**Comment by sharrman** [comment]
I have also added some min max meso per mob lvl. maybe that is also in some lua function? is this related to Data\lua\Precompiled.m2d ?

**Reply by kOchirasu** [reply]
I have decompiled the lua functions `Precompiled.m2d` before, but do not see this function there, so I don't know where it is.

**Comment by sharrman** [comment]
I think I found that function inside luapack in that Precompiled.m2d and also a lot of other functions, some broken but ok...

calcTaxiCharge = function(l_23_0, l_23_1)
  local l_24_2 = 0
  local l_24_3 = {}
  l_24_3.A1 = 0.35307
  l_24_3.B1 = -1.4401
  l_24_3.C1 = 34.075
  l_24_3.A2 = 0.23451
  l_24_3.B2 = 24.221
  l_24_3.C2 = 265.66
  if l_23_1 <= 24 then
    l_24_2 = l_24_3.A1 * l_23_1 ^ 2 + l_24_3.B1 * l_23_1 + l_24_3.C1
  elseif l_23_1 >= 25 then
    local l_24_6 = l_24_3.A2 * l_23_1 - 24 ^ 2
    l_24_6 = l_24_6 + l_24_3.B2 * (l_23_1 - 24)
    l_24_2 = l_24_6 + l_24_3.C2
  end
  local l_24_4 = math.floor
  local l_24_5 = (l_24_2) * l_23_0 / 2 + 0.5
  return l_24_4(l_24_5)
end

**Reply by AngeloTadeucci** [reply]
Readable + Working 

```
calcTaxiCharge = function(maps, level)
    local var1 = 0
    local object = {}
    object.A1 = 0.35307
    object.B1 = -1.4401
    object.C1 = 34.075
    object.A2 = 0.23451
    object.B2 = 24.221
    object.C2 = 265.66

    if level <= 24 then
        var1 = object.A1 * level ^ 2 + object.B1 * level + object.C1
    elseif level >= 25 then
        local temp_var2 = object.A2 * (level - 24) ^ 2
        temp_var2 = temp_var2 + object.B2 * (level - 24)
        var1 = temp_var2 + object.C2
    end

    local result = (var1) * maps / 2 + 0.5
    return math.floor(result)
end
```

**Comment by sharrman** [comment]
calcAirTaxiCharge = function(l_24_0)
  return 30000 + math.max(l_24_0 - 10, 0) * 500
end

**Reply by AngeloTadeucci** [reply]
`l_24_0` it's either level or distance between the maps.

**Reply by sharrman** [reply]
Seems to be level according to client regardless of distance, in kms2 the calculation is diferent.

**Reply by AngeloTadeucci** [reply]
Yeah it's level!

**Comment by AngeloTadeucci** [comment]
- Taxi formula was implemented on PR #332
